drivers/wireless/ieee80211/bcm43xxx/bcmf_chip_43362.c

```diff
@@ -45,39 +45,41 @@ extern const uint8_t bcm43362_firmware_image[];
 extern const unsigned int bcm43362_firmware_image_len;
 #endif
 
-const struct bcmf_sdio_chip bcmf_43362_config_sdio =
+struct bcmf_sdio_chip bcmf_43362_config_sdio;
```

Comment:
it isn't so good to remove const just to conform c89.

Reply:
Why we can't keen `const` and just add missing items to initializer list?

Reply:
Initially I was thinking about replacing {.ram_base = 0, .ram_size = 0x3c000, ... } by { 0,  0x3c000, ... }, however the main problem is with .core_base. There are defined indices for each core. The indices are defined by enum in bcmf_sdio_core.h. The existing approach is auto-synced. If we transform the existing initialization into {0x18000000,  0x18001000, 0x18002000, ...} the auto-sync will not work anymore, that is also dangerous. E.g. someone changes the sequence in the enum, then the referencing code will point out to wrong base addresses.
Possibly it is reasonable then to replace the enum by number of separate defines?
```
#define CHIPCOMMON_CORE_ID  0
#define DOT11MAC_CORE_ID  1
#define SDIOD_CORE_ID  2
...

```
I hope in that case a developer who will modify the code in the future should be aware that the defines are not flexible rather than enum is. 
What do you think?

Reply:
In my opinion breaking the existing useful enum approach (auto-syncing indices) is the worst measure.
Removing const in the structure definition is not good, though it is not very bad compared to breaking enum approach.
Possibly we should give up supporting C89 at least for the bcm43xxx driver.

Reply:
> Initially I was thinking about replacing {.ram_base = 0, .ram_size = 0x3c000, ... } by { 0, 0x3c000, ... }, however the main problem is with .core_base. There are defined indices for each core. The indices are defined by enum in bcmf_sdio_core.h. The existing approach is auto-synced. If we transform the existing initialization into {0x18000000, 0x18001000, 0x18002000, ...} the auto-sync will not work anymore, that is also dangerous. E.g. someone changes the order in the enum, then the referencing code will point out to wrong base addresses. Possibly it is reasonable then to replace the enum by number of separate defines?
> 
> ```
> #define CHIPCOMMON_CORE_ID  0
> #define DOT11MAC_CORE_ID  1
> #define SDIOD_CORE_ID  2
> ...
> ```
> 
> I hope in that case a developer who will modify the code in the future should be aware that the defines are not flexible rather than enum is. What do you think?

Without the compiler support, we can only add the comment in the code to warn other developer.

> In my opinion breaking the existing useful enum approach (auto-syncing indices) is the worst measure. Removing const in the structure definition is not good, though it is not very bad compared to breaking the enum approach. Possibly we should give up supporting C89 at least for the bcm43xxx driver.

Designated initializer is designed to fix the dangerousness you mention, It's unfortunate that c89 doesn't support it. Since only a few arch compiler can't support c99, if bcm43xxx isn't used on these limited arch, why we waste the time to fix the issue and make the code more ugly?

Reply:
I prefer not to follow C89 standard at least in case of bcm43xxx driver (I personally do not need C89 at least now). Though, if we consider the general rule that all the common NuttX code should conform C89, as I understand, the requirement depends not only on arch, however it depends also on a particular MCU (whether it has SDIO host interface). There is a number of architectures that do not support C99. Then for those we should decide if bcm43xxx driver still should follow the C89 requirement. I suppose even if all existing MCUs of the consideration do not have SDIO host interface (it is not clear what the list of the MCUs is), anyway the SDIO host controller can be potentially used outside of the MCU/CPU. SDIO host controller can be some external chip. Or all the parts can be implemented e.g. in FPGA (CPU core + SDIO host controller). E.g. it can be Z80 core (https://opencores.org/projects/t80) + SDIO controller (https://opencores.org/projects/spimaster).

Reply:
Perhaps we could postpone the decision and adapt bcm43xxx driver to C89 later in case of a need. Moreover the original implementation of the driver in NuttX did not follow C89.
So far I could remove C89 related changes from the patch.

Reply:
Yes, this is also my feeling: the core common code(e.g. sched, fs etc) which is always used by any config has to follow c89 to make the old arch happy, but for a specific driver, I think the better approach is waiting that driver is really used by the old arch.